env/ipp_aes_ctr_provider.cc

```diff
@@ -8,136 +8,86 @@
 
 #include "ipp_aes_ctr_provider.h"
 
-#include <regex>
+#include <vector>
+
+#include "util/string_util.h"
 
 #endif
 
 namespace ROCKSDB_NAMESPACE {
 
 #ifndef ROCKSDB_LITE
 
-#if defined(IPPCP) && (defined(HAVE_SSE42) || defined(HAVE_SSE2))
-
-IppAESCTRCipherStream::IppAESCTRCipherStream(IppsAESSpec* aes_ctx,
-                                             const char* init_vector)
-    : aes_ctx_(aes_ctx) {
-  init_vector_ = _mm_loadu_si128((__m128i*)init_vector);
-}
-
-Status IppAESCTRCipherStream::Encrypt(uint64_t fileOffset, char* data,
-                                      size_t dataSize) {
-  if (!dataSize) return Status::OK();
-
-  size_t index = fileOffset / kBlockSize;
-  size_t offset = fileOffset % kBlockSize;
-
-  Ipp8u ctr_block[kBlockSize];
-
-  __m128i counter = _mm_add_epi64(init_vector_, _mm_cvtsi64_si128(index));
-  Ipp8u* ptr_counter = (Ipp8u*)&counter;
-  for (size_t i = 0; i < kBlockSize; ++i)
-    ctr_block[i] = ptr_counter[kBlockSize - 1 - i];
-
-  IppStatus ipp_status = ippStsNoErr;
-  if (!offset) {
-    ipp_status = ippsAESEncryptCTR((Ipp8u*)(data), (Ipp8u*)data, dataSize,
-                                   aes_ctx_, ctr_block, kCounterLen);
-  } else {
-    Ipp8u zero_block[kBlockSize]{0};
-    ipp_status = ippsAESEncryptCTR(zero_block, zero_block, kBlockSize, aes_ctx_,
-                                   ctr_block, kCounterLen);
-    size_t n = std::min(kBlockSize - offset, dataSize);
-    for (size_t i = 0; i < n; ++i) data[i] ^= zero_block[offset + i];
-    memset(zero_block, 0, kBlockSize);
-
-    n = kBlockSize - offset;
-    if (dataSize > n) {
-      Ipp8u* ptr = (Ipp8u*)(data + n);
-      ipp_status = ippsAESEncryptCTR(ptr, ptr, dataSize - n, aes_ctx_,
-                                     ctr_block, kCounterLen);
-    }
+Status IppAESCTRProvider::CreateProvider(
+    const std::string& id, std::shared_ptr<EncryptionProvider>* provider) {
+  if (id != EncryptionProviderNames::kIPPAESProviderName) {
+    return Status::NotSupported("Invalid provider ", id);
   }
-
-  if (ipp_status == ippStsNoErr) return Status::OK();
-
-  return Status::Aborted(ippcpGetStatusString(ipp_status));
-}
-
-Status IppAESCTRCipherStream::Decrypt(uint64_t fileOffset, char* data,
-                                      size_t dataSize) {
-  return Encrypt(fileOffset, data, dataSize);
+  IppAESCTRProvider* ptr_ipp = new IppAESCTRProvider();
+  provider->reset(ptr_ipp);
+  return Status::OK();
 }
 
-#endif  // IPPCP
-
-Status IppAESCTRProvider::CreateProvider(const std::string& id,
-                                         IppAESCTRProvider** provider) {
+Status IppAESCTRProvider::AddCipher(const std::string& /*descriptor*/,
+                                    const char* cipher, size_t len,
+                                    bool /*for_write*/) {
```

Comment:
You either need:
- Support more than one encryption key
- Prevent the user from calling this method if there is already an encryption key (return an error)
- If this method is called and there already is a key, delete the old key before you insert the new one